sui-execution/latest/sui-adapter/src/temporary_store.rs

```diff
@@ -607,113 +633,121 @@ impl<'backing> TemporaryStore<'backing> {
+                        // failure), whereas adding the immutable object to the roots will prevent
+                        // us from catching this.
+                        None
+                    }
+                    Owner::ObjectOwner(_parent) => {
+                        unreachable!("Input objects must be address owned, shared, or immutable")
+                    }
                 }
-            }
-        }
-        Ok((objs_to_authenticate, authenticated_objs))
-    }
-
-    // check that every object read is owned directly or indirectly by sender, sponsor, or a shared object input
-    pub fn check_ownership_invariants(
-        &self,
-        sender: &SuiAddress,
-        gas_charger: &mut GasCharger,
-        is_epoch_change: bool,
-    ) -> SuiResult<()> {
-        let (mut objects_to_authenticate, mut authenticated_objects) =
-            self.get_objects_to_authenticate(sender, gas_charger, is_epoch_change)?;
+            })
+            .filter(|id| {
+                // remove any non-mutable inputs. This will remove deleted or readonly shared
+                // objects
+                mutable_inputs.contains(id)
+            })
+            .copied()
+            .collect();
 
+        // check all modified objects are authenticated (excluding gas objects)
+        let mut objects_to_authenticate = self
+            .execution_results
+            .modified_objects
+            .iter()
+            .filter(|id| !gas_objs.contains(id))
+            .copied()
+            .collect::<Vec<_>>();
         // Map from an ObjectID to the ObjectID that covers it.
-        let mut covered = BTreeMap::new();
         while let Some(to_authenticate) = objects_to_authenticate.pop() {
-            let Some(old_obj) = self.store.get_object(&to_authenticate)? else {
-                // lookup failure is expected when the parent is an "object-less" UID (e.g., the ID of a table or bag)
-                // we cannot distinguish this case from an actual authentication failure, so continue
+            if authenticated_for_mutation.contains(&to_authenticate) {
+                // object has been authenticated
                 continue;
-            };
-            let parent = match &old_obj.owner {
-                Owner::ObjectOwner(parent) | Owner::AddressOwner(parent) => ObjectID::from(*parent),
-                owner => panic!(
-                    "Unauthenticated root at {to_authenticate:?} with owner {owner:?}\n\
-             Potentially covering objects in: {covered:#?}",
-                ),
-            };
-
-            if authenticated_objects.contains(&parent) {
-                authenticated_objects.insert(to_authenticate);
-            } else if !covered.contains_key(&parent) {
-                objects_to_authenticate.push(parent);
             }
-
-            covered.insert(to_authenticate, parent);
+            let wrapped_parent = self.wrapped_object_containers.get(&to_authenticate);
```

Comment:
Is this just an optimization to avoid unnecessary object store loads? Not sure I fully understand reading from the wrapped object containers

Reply:
It's necessary for the expensive invariant checks. It just improves our precision. 
Previously, we just assumed that any wrapped ID was authenticated. Now we will check that the wrapped object's container is authenticated.

Reply:
Still a bit confused. So we are iterating through the `modified_objects` list, which can only contain objects that actually exist in the object store before we commit the effects of this transaction. Due to this, I would expect that these objects can never show up in wrapped_object_containers?